plugins/woocommerce-admin/client/customize-store/index.tsx

```diff
@@ -0,0 +1,193 @@
+	predictableActionArguments: true,
+	preserveActionOrder: true,
+	schema: {
+		context: {} as customizeStoreStateMachineContext,
+		events: {} as customizeStoreStateMachineEvents,
+		services: {} as {
+			fetchThemeCards: { data: ThemeCard[] };
+		},
+	},
+	context: {
+		intro: {
+			themeCards: [] as ThemeCard[],
+			activeTheme: '',
+		},
+	} as customizeStoreStateMachineContext,
+	states: {
+		intro: {
+			id: 'intro',
+			initial: 'preIntro',
+			states: {
+				preIntro: {
+					invoke: {
+						src: 'fetchThemeCards',
+						onDone: {
+							target: 'intro',
+							actions: [ 'assignThemeCards' ],
+						},
+					},
+				},
+				intro: {
+					meta: {
+						component: Intro,
+					},
+				},
+			},
+			on: {
+				DESIGN_WITH_AI: {
+					target: 'designWithAi',
+				},
+				SELECTED_ACTIVE_THEME: {
+					target: 'assemblerHub',
+				},
+				CLICKED_ON_BREADCRUMB: {
+					target: 'backToHomescreen',
+				},
+				SELECTED_NEW_THEME: {
+					target: '? Appearance Task ?',
+				},
+				SELECTED_BROWSE_ALL_THEMES: {
+					target: '? Appearance Task ?',
+				},
+			},
+		},
+		designWithAi: {
+			initial: 'preDesignWithAi',
+			states: {
+				preDesignWithAi: {
+					always: {
+						target: 'designWithAi',
+					},
+				},
+				designWithAi: {
+					meta: {
+						component: DesignWithAi,
+					},
+				},
+			},
+			on: {
+				THEME_SUGGESTED: {
+					target: 'assemblerHub',
+				},
+			},
+		},
+		assemblerHub: {
+			on: {
+				FINISH_CUSTOMIZATION: {
+					target: 'backToHomescreen',
+				},
+			},
+		},
+		backToHomescreen: {},
+		'? Appearance Task ?': {},
+	},
+} );
+
+export const CustomizeStoreController = ( {
+	actionOverrides,
+	servicesOverrides,
+}: {
+	actionOverrides: Partial< typeof customizeStoreStateMachineActions >;
+	servicesOverrides: Partial< typeof customizeStoreStateMachineServices >;
+} ) => {
+	useFullScreen( [ 'woocommerce-customize-store' ] );
+
+	const augmentedStateMachine = useMemo( () => {
+		return customizeStoreStateMachineDefinition.withConfig( {
+			services: {
+				...customizeStoreStateMachineServices,
+				...servicesOverrides,
+			},
+			actions: {
+				...customizeStoreStateMachineActions,
+				...actionOverrides,
+			},
+			guards: {},
+		} );
+	}, [ actionOverrides, servicesOverrides ] );
+
+	const [ state, send, service ] = useMachine( augmentedStateMachine, {
+		devTools: process.env.NODE_ENV === 'development',
+	} );
+
+	// eslint-disable-next-line react-hooks/exhaustive-deps -- false positive due to function name match, this isn't from react std lib
+	const currentNodeMeta = useSelector( service, ( currentState ) =>
+		findComponentMeta< CustomizeStoreComponentMeta >(
+			currentState?.meta ?? undefined
+		)
+	);
+
+	const [ CurrentComponent, setCurrentComponent ] =
+		useState< CustomizeStoreComponent | null >( null );
+	useEffect( () => {
+		if ( currentNodeMeta?.component ) {
+			setCurrentComponent( () => currentNodeMeta?.component );
+		}
+	}, [ CurrentComponent, currentNodeMeta?.component ] );
```

Comment:
I know its copied code but I think this reduces chances for unnecessary re-renders.
```suggestion
	useEffect( () => {
		if (
			currentNodeMeta?.component &&
			currentNodeMeta?.component !== CurrentComponent
		) {
			setCurrentComponent( () => currentNodeMeta?.component );
		}
	}, [ CurrentComponent, currentNodeMeta?.component ] );
```

Reply:
I think the suggested code is wrong 🙈 
CurrentComponent is set to `() => currentNodeMeta.component` so testing them for inequality will always be true

Additionally their existence in the useEffect dependency array guarantees that the code is only run in event of referential inequality with their previous values, so it's probably not necessary to test again

Reply:
Ah, right. Nevermind this 👍